common/src/main/java/com/wynntils/wynn/model/bossbar/BossBarModel.java

```diff
@@ -0,0 +1,219 @@
+
+        public void add(
+                UUID id,
+                Component name,
+                float progress,
+                BossEvent.BossBarColor color,
+                BossEvent.BossBarOverlay overlay,
+                boolean darkenScreen,
+                boolean playMusic,
+                boolean createWorldFog) {
+            TrackedBar trackedBar = null;
+            Matcher matcher = null;
+
+            // TODO when we can successfuly parse character info, reduce checks
+            for (TrackedBar potentialTrackedBar :
+                    Arrays.asList(manaBankBar, bloodPoolBar, awakenedBar, focusBar, corruptedBar)) {
+                matcher = potentialTrackedBar.pattern.matcher(ComponentUtils.getCoded(name));
+                if (matcher.matches()) {
+                    trackedBar = potentialTrackedBar;
+                    break;
+                }
+            }
+
+            if (trackedBar == null) return;
+
+            trackedBar.setProgress(progress);
+            trackedBar.setUuid(id);
+
+            // Allow for others to try and cancel event
+            CustomBarAddEvent barAddEvent = new CustomBarAddEvent(trackedBar.type);
+            WynntilsMod.postEvent(barAddEvent);
+
+            if (barAddEvent.isCanceled()) {
+                trackedBar.setRendered(false);
+                event.setCanceled(true);
+            } else {
+                trackedBar.setRendered(true);
+            }
+
+            trackedBar.onUpdateName(matcher);
+
+            trackedBarsMap.put(id, trackedBar);
+        }
+
+        private void whenBarPresent(UUID id, Consumer<TrackedBar> consumer) {
+            TrackedBar trackedBar = trackedBarsMap.get(id);
+
+            if (trackedBar != null) {
+                if (!trackedBar.isRendered()) {
+                    event.setCanceled(true);
+                }
+
+                consumer.accept(trackedBar);
+            }
+        }
+
+        public void remove(UUID key) {
+            whenBarPresent(key, trackedBar -> {
+                trackedBar.reset();
+                trackedBarsMap.remove(key);
+            });
+        }
+
+        public void updateProgress(UUID id, float progress) {
+            whenBarPresent(id, trackedBar -> {
+                trackedBar.setProgress(progress);
+            });
+        }
+
+        public void updateName(UUID id, Component name) {
+            whenBarPresent(id, trackedBar -> {
+                Matcher matcher = trackedBar.pattern.matcher(ComponentUtils.getCoded(name));
+                if (!matcher.matches()) {
+                    WynntilsMod.error("Failed to match already matched boss bar");
+                }
+
+                trackedBar.onUpdateName(matcher);
+            });
+        }
+
+        public void updateStyle(UUID id, BossEvent.BossBarColor color, BossEvent.BossBarOverlay overlay) {
+            whenBarPresent(id, trackedBar -> {});
```

Comment:
Is this needed to get the `event.setCanceled(true);`? Otherwise this seems like a no-op. If my guess is correct then perhaps a comment explaining this side-effect of whenBarPresent. Or maybe even better rename it to `handleBossBarUpdate`; this does not indicate that the sole purpose is to run a lambda if the bar is present, which the current name does.